### ISSUE TITLE
X.509 certificate Issuer and Subject not checked

### ISSUE THREAD

**Issue by JasonVranek** [PufferFinance/rave#12]
The `verifySignedX509()` function is used to check whether an X.509 certificate is valid. The certificates are assumed to be issued by the Intel Attestation Service. According to the IAS documentation, the signing chain will have certificates matching two specific distinguished names:

1. CN=Intel SGX Attestation Report Signing CA, O=Intel Corporation, L=Santa Clara, ST=CA, C=US
2. CN=Intel SGX Attestation Report Signing, O=Intel Corporation, L=Santa Clara, ST=CA, C=US

The second distinguished name corresponds to issuer of the `cert` argument to `verifySignedX509()`. However, the function does not check that the `Subject` (if it exists) and `Issuer`'s respective distinguished names matches the ones described in the IAS documentation.

Note that [[IETF RFC 5280](https://datatracker.ietf.org/doc/html/rfc5280#section-4.1.2.4)](https://datatracker.ietf.org/doc/html/rfc5280#section-4.1.2.4) section 4.1.2.4 also notes the following about the `Issuer` field:

> Certificate users MUST be prepared to process the issuer
distinguished name and subject distinguished name (Section 4.1.2.6) fields to perform name chaining for certification path validation (Section 6). Name chaining is performed by matching the issuer distinguished name in one certificate with the subject name in a CA certificate.
> 

## Impact

By omitting checks for the `Subject` and `Issuer`, it will be easier for an attacker to provide a fraudulent signing certificate in the event of a breach. For example, if the smart contracts interfacing with the RAVe smart contracts are compromised, and an attacker is able to supply an attacker-controlled parent certificate (e.g., such as a self-signed certificate), then the attacker can supply any signing certificate, thereby allowing the attacker to arbitrarily forge signatures with RAVe.

In contrast, with such checks in place, the attacker will be forced to provide both a parent and a signing certificate whose respective distinguished names must match the ones specified in the IAS documentation. This improves the security of the system by making such an attack slightly more difficult. Additionally, such checks can also help guard against configuration errors.

## Recommendation

The `verifySignedX509()` function should check that the attributes in the `Issuer` and `Subject` match the ones in the IAS documentation. This also helps enforce specific structures on the certificate that could help prevent attacks that exploit [[getPtr does not handle out-of-bounds integers](https://www.notion.so/getPtr-does-not-handle-out-of-bounds-integers-6ec6978b62bc49b49004f9e1c6193fe4?pvs=21)](https://www.notion.so/getPtr-does-not-handle-out-of-bounds-integers-6ec6978b62bc49b49004f9e1c6193fe4?pvs=21).

As an additional layer of security, the developers could also implement some basic “name chaining” and validate the `Issuer` and `Subject` of the parent certificate as well.

**Comment by robertsdotpm** [comment]
Maybe I misunderstand but the biggest flaw to me is that all of the RSA parameters are dynamically input to the RAVE function. Like the root CA signer mod and exp. The checks are irrelevant if someone can just generate a self-signed cert and child cert, pass in their own parameters and pass the check. I think if we fix the contract to use hard-coded values for the root cert CA then the issue here is solved as a side-effect without having to verify name chaining. We can still check the subject of leaf certs matches the expected value. But with this change attackers can't just generate their own self-signed certs.

**Comment by JasonVranek** [comment]
I'm in favor, especially since the root cert has a long expiry date

**Comment by robertsdotpm** [comment]
Okay, I've added checks for subject and issuer. There is now hard-coded values for Intels report signer CA exp and mod. Setting empty bytes makes the contract use these values. Otherwise, functions can pass their own values for tests and so fourth. The x509 verification code disables checks for subject and issuer if the passed in params don't match Intels. It's currently working. I have noted that the 'rave sanity' test code is still flaky but this is an issue with the test code and not with RAVE. Closing this issue for now.
